tick/optim/solver/src/svrg.cpp

```diff
@@ -1,127 +1,150 @@
+  }
+  ready_step_corrections = true;
+}
 
-    if (variance_reduction == VarianceReductionMethod::Random ||
-        variance_reduction == VarianceReductionMethod::Average) {
-        next_iterate.init_to_zero();
+void SVRG::solve_dense() {
+  for (ulong t = 0; t < epoch_size; ++t) {
+    ulong i = get_next_i();
+    model->grad_i(i, iterate, grad_i);
+    model->grad_i(i, fixed_w, grad_i_fixed_w);
+    for (ulong j = 0; j < iterate.size(); ++j) {
+      iterate[j] = iterate[j] - step * (grad_i[j] - grad_i_fixed_w[j] + full_gradient[j]);
     }
-
-    if (variance_reduction == VarianceReductionMethod::Random) {
-        rand_index = rand_unif(epoch_size);
+    prox->call(iterate, step, iterate);
+    if (variance_reduction == VarianceReductionMethod::Random && t == rand_index) {
+      next_iterate = iterate;
     }
-
-    for (ulong t = 0; t < epoch_size; ++t) {
-        ulong i = get_next_i();
-        // Sparse features vector
-        BaseArrayDouble x_i = model->get_features(i);
-        // Gradients factor
-        double alpha_i_iterate = model->grad_i_factor(i, iterate);
-        double alpha_i_fixed_w = model->grad_i_factor(i, fixed_w);
-        double delta = -step * (alpha_i_iterate - alpha_i_fixed_w);
-        if (use_intercept) {
-            // Get the features vector, which is sparse here
-            ArrayDouble iterate_no_interc = view(iterate, 0, n_features);
-            //
-            iterate_no_interc.mult_incr(x_i, delta);
-            iterate[n_features] += delta;
-            iterate.mult_incr(mu, -step);
-        } else {
-            iterate.mult_incr(x_i, delta);
-            iterate.mult_incr(mu, -step);
-        }
-
-        prox->call(iterate, step, iterate);
-
-        if (variance_reduction == VarianceReductionMethod::Random && t == rand_index)
-            next_iterate = iterate;
-
-        if (variance_reduction == VarianceReductionMethod::Average)
-            next_iterate.mult_incr(iterate, 1.0 / epoch_size);
+    if (variance_reduction == VarianceReductionMethod::Average) {
+      next_iterate.mult_incr(iterate, 1.0 / epoch_size);
     }
+  }
+  if (variance_reduction == VarianceReductionMethod::Last) {
+    next_iterate = iterate;
+  }
+  t += epoch_size;
+}
 
-    if (variance_reduction == VarianceReductionMethod::Last)
-        next_iterate = iterate;
+void SVRG::solve_sparse_proba_updates(bool use_intercept, ulong n_features) {
+  // Data is sparse, and we use the probabilistic update strategy
+  // This means that the model is a child of ModelGeneralizedLinear.
+  // The strategy used here uses non-delayed updates, with corrected
+  // step-sizes using a probabilistic approximation and the
+  // penalization trick: with such a model and prox, we can work only inside the current
+  // support (non-zero values) of the sampled vector of features
+  std::shared_ptr<ProxSeparable> casted_prox;
+  if (prox->is_separable()) {
+    casted_prox = std::static_pointer_cast<ProxSeparable>(prox);
+  }
```

Comment:
else raise an error

Otherwise we will have a segfault (sorry I should have added it)

Reply:
Done